std/array.d

```diff
@@ -3927,3 +3927,215 @@ unittest
     assert(appS.data == "hellow");
     assert(appA.data == "hellow");
 }
+
+/++
+Constructs a static array from `a`.
+The type of elements can be specified implicitly (`[1,2].asStatic` of type int[2])
+or explicitly (`[1,2].asStatic!float` of type float[2]).
+When `a` is a range (not known at compile time), the number of elements has to be given as template argument
+(eg `myrange.asStatic!2`).
+Size and type can be combined (eg: `2.iota.asStatic!(byte[2])`).
+When the range `a` is known at compile time, it can also be specified as a
+template argument to avoid having to specify the number of elements
+(eg: `asStatic!(2.iota)` or `asStatic!(double, 2.iota)`).
+
+Note: `foo([1, 2, 3].asStatic)` may be inefficient because of the copies involved.
+
+Params: a = The input elements
+
+Returns: A static array constructed from `a`.
++/
+pragma(inline, true) T[n] asStatic(T, size_t n)(auto ref T[n] a) nothrow @safe pure @nogc
+{
+    return a;
+}
+
+/// ditto
+U[n] asStatic(U, T, size_t n)(auto ref T[n] a)
+if (!is(U == T))
+{
+    import std.conv : emplaceRef;
+    U[n] ret = void;
```

Comment:
This function has a pernicious bug: if construction throws an exception, `ret` will be destroyed and destructors will be called for uninitialized objects. Recall that statically-sized arrays on the stack insert one destructor call for each member.

Reply:
avoided this bug by making these overloads as `nothrow`, so a throwing ctor will cause CT error instead of potential runtime bug.

support for throwing constructors can always be addressed in a future PR (without breaking backward compatibility), maybe something along these lines:

```
U[n] staticArray(U, T, size_t n)(auto ref T[n] a)
if (!is(U == T))
{
     import std.conv : emplaceRef;

     U[n] ret = void;
    size_t i;
    try
     {
        for (; i < n; i++)
        {
            emplaceRef!U(ret[i], cast(U) a[i]);
        }
    }
    catch(Exception t) // catch(Throwable t) causes errors with D linter; but we really want Throwable here for correctness
    {
        for (; i < n; i++)
        {
            emplaceRef!U(ret[i], U.init);
        }
        throw t;
     }
     return ret;
 }
```